fs/metrics_sample.h

```diff
@@ -0,0 +1,208 @@
+        {ZENFS_IO_ALLOC_WAL_ACTUAL_LATENCY, ZENFS_REPORTER_TYPE_LATENCY},
+        {ZENFS_IO_ALLOC_NON_WAL_ACTUAL_LATENCY, ZENFS_REPORTER_TYPE_LATENCY},
+        {ZENFS_META_ALLOC_LATENCY, ZENFS_REPORTER_TYPE_LATENCY},
+        {ZENFS_METADATA_SYNC_LATENCY, ZENFS_REPORTER_TYPE_LATENCY},
+        {ZENFS_ROLL_LATENCY, ZENFS_REPORTER_TYPE_LATENCY},
+        {ZENFS_WRITE_QPS, ZENFS_REPORTER_TYPE_QPS},
+        {ZENFS_READ_QPS, ZENFS_REPORTER_TYPE_QPS},
+        {ZENFS_SYNC_QPS, ZENFS_REPORTER_TYPE_QPS},
+        {ZENFS_IO_ALLOC_QPS, ZENFS_REPORTER_TYPE_QPS},
+        {ZENFS_META_ALLOC_QPS, ZENFS_REPORTER_TYPE_QPS},
+        {ZENFS_ROLL_QPS, ZENFS_REPORTER_TYPE_QPS},
+        {ZENFS_WRITE_THROUGHPUT, ZENFS_REPORTER_TYPE_THROUGHPUT},
+        {ZENFS_ROLL_THROUGHPUT, ZENFS_REPORTER_TYPE_THROUGHPUT},
+        {ZENFS_ACTIVE_ZONES, ZENFS_REPORTER_TYPE_GENERAL},
+        {ZENFS_OPEN_ZONES, ZENFS_REPORTER_TYPE_GENERAL},
+        {ZENFS_FREE_SPACE, ZENFS_REPORTER_TYPE_GENERAL},
+        {ZENFS_USED_SPACE, ZENFS_REPORTER_TYPE_GENERAL},
+        {ZENFS_RECLAIMABLE_SPACE, ZENFS_REPORTER_TYPE_GENERAL},
+        {ZENFS_RESETABLE_ZONES, ZENFS_REPORTER_TYPE_GENERAL}};
+
+struct ReporterSample {
+ public:
+  typedef uint64_t TypeTime;
+  typedef uint64_t TypeValue;
+  typedef std::pair<TypeTime, TypeValue> TypeRecord;
+
+ private:
+  port::Mutex mu_;
+  ZenFSMetricsReporterType type_;
+  std::vector<TypeRecord> hist_;
+
+  static const TypeTime MinReportInterval =
+      30 * 1000000;  // 30 seconds for all reporters by default.
+  bool ReadyToReport(uint64_t time) const {
+    // AssertHeld(&mu);
+    if (hist_.size() == 0) return 1;
+    TypeTime last_report_time = hist_.rbegin()->first;
+    return time > last_report_time + MinReportInterval;
+  }
+
+ public:
+  ReporterSample(ZenFSMetricsReporterType type) : mu_(), type_(type), hist_() {}
+  void Record(const TypeTime& time, TypeValue value) {
+    MutexLock guard(&mu_);
+    if (ReadyToReport(time)) hist_.push_back(TypeRecord(time, value));
+  }
+  ZenFSMetricsReporterType Type() const { return type_; }
+  void GetHistSnapshot(std::vector<TypeRecord>& hist) {
+    MutexLock guard(&mu_);
+    hist = hist_;
+  }
+};
+
+struct ZenFSMetricsSample : public ZenFSMetrics {
+ public:
+  typedef uint64_t TypeMicroSec;
+  typedef ReporterSample TypeReporter;
+
+ private:
+  Env* env_;
+  std::unordered_map<ZenFSMetricsHistograms, TypeReporter> reporter_map_;
+
+ public:
+  ZenFSMetricsSample(Env* env) : env_(env), reporter_map_() {
+    for (auto& label_with_type : ZenFSHistogramsTypeMap)
+      AddReporter(static_cast<uint32_t>(label_with_type.first),
+                  static_cast<uint32_t>(label_with_type.second));
+  }
+  ~ZenFSMetricsSample() {}
+
+  virtual void AddReporter(uint32_t label_uint,
+                           uint32_t type_uint = 0) override {
+    auto label = static_cast<ZenFSMetricsHistograms>(label_uint);
+    assert(ZenFSHistogramsNameMap.find(label) !=
+           BDZenFSHistogramsNameMap.end());
```

Comment:
This looks like a typo

Reply:
yes.
I may have compiled it without saving it, so I didn't catch the error.